packages/hardhat-core/src/internal/vendor/await-semaphore/index.ts

```diff
@@ -64,3 +67,154 @@ export class Mutex extends Semaphore {
+  // The _safeMarginInMs is employed here to ensure that waiting processes do not prematurely terminate the current mutex holder
+  // before it reports an error for exceeding the allowed execution time.
+  private _log: debug.Debugger;
+  private _mutexFilePath: string;
+  private _timeoutInMs: number;
+  private _defaultTimeoutInMs = 20000;
+  private _safeMarginInMs = 2000;
+  private _mutexReleaseWaitingTimeInMs = 500;
+
+  constructor(mutexName: string, timeoutInMs?: number) {
+    this._log = debug("hardhat:await-semaphore:multi-process-mutex");
+
+    this._log(`Creating mutex with name '${mutexName}'`);
+
+    this._mutexFilePath = `${os.tmpdir()}/${mutexName}.txt`;
+    this._timeoutInMs = timeoutInMs ?? this._defaultTimeoutInMs;
+  }
+
+  public async use<T>(f: () => Promise<T>): Promise<T> {
+    this._log(
+      `Starting mutex process withy mutex file '${this._mutexFilePath}'`
+    );
+
+    while (true) {
+      if (await this._tryToAcquireMutex()) {
+        // Mutex has been acquired
+        return this._executeFunctionAndReleaseMutex(f);
+      }
+
+      // Mutex not acquired
+      if (this._isMutexFileTooOld()) {
+        // If the mutex file is too old, it likely indicates a deadlock, so the file should be removed
+        this._log(
+          `Current mutex file is too old, removing it at path '${this._mutexFilePath}'`
+        );
+        this._deleteMutexFile();
+      } else {
+        // wait
+        await this._waitMs();
+      }
+    }
+  }
+
+  private async _tryToAcquireMutex() {
+    try {
+      // Create a file only if it does not exist
+      fs.writeFileSync(this._mutexFilePath, "", { flag: "wx+" });
+      return true;
+    } catch (error: any) {
+      if (error.code === "EEXIST") {
+        // File already exists, so the mutex is already acquired
+        return false;
+      }
+
+      throw error;
+    }
+  }
+
+  private async _executeFunctionAndReleaseMutex<T>(
+    f: () => Promise<T>
+  ): Promise<T> {
+    this._log(`Mutex acquired at path '${this._mutexFilePath}'`);
+
+    // Set a timeout to ensure that the function does not take longer than the allowed time
+    let timeoutHandle: NodeJS.Timeout;
+    const timeoutChecker = new Promise<never>((_, reject) => {
+      timeoutHandle = setTimeout(() => {
+        reject(
+          new Error(
+            "The function's execution timed out while it had ownership of the mutex"
+          )
+        );
+      }, this._defaultTimeoutInMs);
+    });
+
+    try {
+      // Execute the function passed to the mutex, no other function will be able to run at the same time.
+      // If the function will take more than the timeout, it will be terminated with an error and the mutex will be released.
+      // Promise.race is used to race the timeoutChecker against the execution of fn().
+      // Whichever completes first will be the outcome of Promise.race.
+      return await Promise.race([f(), timeoutChecker]).then((result) => {
```

Comment:
Is there a reason to use `.then` here instead of just doing `const result = await Promise.race(...)`?

Reply:
This logic has been removed, see previous comments